File: soungmunkim/Python/Sorting/HeapSort.py
def heapify(arr, n, i):
    largest = i  # 현재 노드
    l = 2 * i + 1  # 왼쪽 자식 노드
    r = 2 * i + 2  # 오른쪽 자식 노드

    # 왼쪽 자식 노드가 존재하며, 현재 노드보다 크다면
    if l < n and arr[i] < arr[l]:
        largest = l

    # 오른쪽 자식 노드가 존재하며, largest보다 크다면
    if r < n and arr[largest] < arr[r]:
        largest = r

    # largest가 현재 노드가 아니라면
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # 노드 교환

        heapify(arr, n, largest)  # 재귀적으로 힙 구성

def heapSort(arr):
    n = len(arr)

    # max heap을 구성
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # 원소 하나씩 추출하면서 힙 정렬
    for i in range(n-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # 현재 루트와 마지막 원소 교환
        heapify(arr, i, 0)  # 루트를 제외하고 다시 힙 구성

    return arr

arr = [9, 6, 7, 5, 3, 1, 4, 2]
print("Original array:", arr)
sorted_arr = heapSort(arr)
print("Sorted array:", sorted_arr)

File: soungmunkim/C/Sorting/HeapSort.c
#include <stdio.h>

// 배열을 최대 힙 구조로 만드는 함수
void heapify(int *arr, int size) {
    for (int i = 1; i < size; ++i) {
        int child = i;
        do {
            // 자식 노드가 부모 노드보다 크면 교환
            int root = (child - 1) / 2;
            if (arr[root] < arr[child]) {
                // 부모 노드와 자식 노드의 값을 교환
                int temp = arr[root];
                arr[root] = arr[child];
                arr[child] = temp;
            }
            child = root;
        } while (child != 0); // 최상단 노드까지 점검
    }
}

// 최상단 노드와 최하단 노드를 교체하는 함수
void heap(int *arr, int *size) {
    int temp = arr[0];
    arr[0] = arr[*size - 1];
    arr[*size - 1] = temp;
    --(*size);
}

int main() {

    int size = 10;
    // 무작위 배열
    int arr[10] = {5, 6, 10, 4, 3, 8, 7, 1, 2, 9};
    printf("원래 배열: ");
    for (int i = 0; i < 10; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    // 힙 정렬
    printf("정렬 후 배열: ");
    for (int i = 0; i < 10; ++i) {
        heapify(arr, size); // 배열을 최대 힙 구조로 만듦
        heap(arr, &size);   // 최대 힙에서 최대값을 빼고 배열 크기를 줄임
    }

    // 정렬된 배열 출력
    for (int i = 0; i < 10; ++i) {
        printf("%d ", arr[i]);
    }

    return 0;
}
